File: verifiers/draw_verify.py
#!/usr/bin/env python3
# ============================================================================
# UVS verifiable allocation - REFERENCE VERIFIER (Python 3, stdlib only)
#
# One operation: a seeded random PERMUTATION of participants, then a published
# prize pool dealt onto that order.
#   combinedSeed = SHA-256( serverSeed + ":" + drandRandomness )
#   score(id)    = SHA-256( combinedSeed + ":" + id )
#   permutation  = participants sorted by score DESC  (ties: id ASC)
#   allocation   = order[i] receives prizes[i]
# Same result as draw-verify.js / DrawVerify.java / draw_verify.cpp.
#
# Run:  python draw_verify.py <record.json> [id]
# ============================================================================
import hashlib, json, sys
from functools import cmp_to_key

def sha256(s): return hashlib.sha256(s.encode('utf-8')).hexdigest()
def combined_seed(server, rand): return sha256(server + ':' + rand)
def score_of(combined, idv): return sha256(combined + ':' + idv)

def require_unique(parts):   # uvLs 3.1: duplicate ids break the total order -> reject, don't rank
    if len(set(parts)) != len(parts):
        raise ValueError('INVALID: duplicate participant ids - record rejected (uvLs 3.1)')

def _cmp(a, b):           # a,b = (score, id); score DESC, id ASC
    if a[0] > b[0]: return -1
    if a[0] < b[0]: return 1
    if a[1] < b[1]: return -1
    if a[1] > b[1]: return 1
    return 0

def permute(parts, combined):
    require_unique(parts)
    return sorted([(score_of(combined, i), i) for i in parts], key=cmp_to_key(_cmp))

def allocate(parts, combined, prizes):
    order = permute(parts, combined)
    return [{'id': p[1], 'rank': i + 1, 'prize': prizes[i] if i < len(prizes) else None} for i, p in enumerate(order)]

def lookup(parts, combined, idv, prizes):
    require_unique(parts)
    me = score_of(combined, idv)
    higher, present = 0, False
    for a in parts:
        if a == idv:
            present = True; continue
        s = score_of(combined, a)
        if s > me or (s == me and a < idv): higher += 1
    rank = higher + 1
    return {'id': idv, 'present': present, 'rank': rank,
            'prize': prizes[rank - 1] if present and rank <= len(prizes) else None}

# §6.1 proportional pools: a tier count derived from M as an integer num/den with one rounding
# mode. All-integer so JS (BigInt), Python, Java (long) and C++ (int64) agree; operands are
# non-negative, so integer division is floor.
def resolve_count(M, rule):
    num, den, mode = rule.get('num'), rule.get('den'), rule.get('mode', 'round-half-up')
    ints = all(isinstance(x, int) and not isinstance(x, bool) for x in (num, den, M))
    if not ints or den <= 0 or num < 0 or M < 0:
        raise ValueError('INVALID: proportional num/den/M must be non-negative integers with den>0 (uvLs 6.1)')
    if mode == 'floor': return (M * num) // den
    if mode == 'ceil':  return (M * num + den - 1) // den
    if mode == 'round-half-up': return (2 * M * num + den) // (2 * den)
    raise ValueError('INVALID: unknown rounding mode "%s" (uvLs 6.1)' % mode)

def pool_of(rec):
    if isinstance(rec.get('prizes'), list): return rec['prizes']
    r = rec.get('rules') or {}
    if isinstance(r.get('prizes'), list): return r['prizes']
    # a published draw record nests the pool under `rules` (uvLs 7); accept either shape.
    pp = rec['prizePool'] if isinstance(rec.get('prizePool'), list) else (r['prizePool'] if isinstance(r.get('prizePool'), list) else None)
    if pp is not None:
        M, prizes, total = len(rec['participants']), [], 0
        for t in pp:
            if not isinstance(t.get('tier'), str):
                raise ValueError('INVALID: tier label must be a string (uvLs 6)')
            if t.get('rule') is not None:
                count = resolve_count(M, t['rule'])
                if t.get('count') is not None and t['count'] != count:   # §6.1/§9.4 mismatch -> reject
                    raise ValueError('INVALID: tier "%s" count %s != rule-resolved %d (uvLs 6.1)' % (t['tier'], t['count'], count))
            else:
                if not isinstance(t.get('count'), int) or isinstance(t.get('count'), bool):
                    raise ValueError('INVALID: tier count must be a JSON integer (uvLs 6.1)')
                count = t['count']
            if total + count > M: count = M - total      # §6.1 ordering: clamp running total to M
            prizes.extend([t['tier']] * count)
            total += count
        return prizes
    n = rec.get('winners') or rec.get('N') or r.get('winners') or r.get('N') or 0
    return [rec.get('prizeLabel') or r.get('prizeLabel') or 'WIN'] * n

# -- §5.4 anchor round rule (optional) --
# drand quicknet: 3s period, genesis 1692803367. The DERIVED-R rule (uvLs 5.4.1) sets
# R = round_at(genTime)+1, so genTime < time_of_round(R) holds by construction and R is not the
# operator's choice. Confirm ORDERING here; verify the token itself with `openssl ts -verify`.
QN_GENESIS, QN_PERIOD = 1692803367, 3
def round_at(unix_sec): return (unix_sec - QN_GENESIS) // QN_PERIOD + 1
def time_of_round(rnd): return QN_GENESIS + (rnd - 1) * QN_PERIOD
def check_anchor_round(gen_time, rnd):
    expected = round_at(gen_time) + 1
    rt = time_of_round(rnd)
    return {'ok': rnd == expected and gen_time < rt, 'expectedRound': expected, 'roundTime': rt, 'genBeforeRound': gen_time < rt}

if __name__ == '__main__':
    if len(sys.argv) < 2:
        print('usage: python draw_verify.py <record.json> [id]'); sys.exit(2)
    rec = json.load(open(sys.argv[1], encoding='utf-8'))
    try: require_unique(rec['participants'])
    except ValueError as e: print(e); sys.exit(1)
    combined = combined_seed(rec['serverSeed'], rec['drand']['randomness'])
    prizes = pool_of(rec)
    print('combinedSeed = SHA-256(serverSeed:drandRandomness) =', combined)
    ca = rec.get('commitmentAnchor')
    if ca and ca.get('genTime') is not None and rec.get('drand', {}).get('round') is not None and str(ca.get('roundRule', '')).startswith('roundAt'):
        c = check_anchor_round(ca['genTime'], rec['drand']['round'])
        print('5.4 derived-R: R==round_at(genTime)+1 ?', rec['drand']['round'] == c['expectedRound'],
              '| genTime<time_of_round(R) ?', c['genBeforeRound'], '->', 'OK' if c['ok'] else 'FAIL')
    if len(sys.argv) > 2:
        r = lookup(rec['participants'], combined, sys.argv[2], prizes)
        if r['present']:
            print('%s: rank %d of %d -> %s' % (r['id'], r['rank'], len(rec['participants']), r['prize'] or 'no prize'))
        else:
            print('%s: NOT in the committed list' % r['id'])
    else:
        a = allocate(rec['participants'], combined, prizes)
        winners = [x for x in a if x['prize'] is not None]
        print('%d prize(s) among %d participants:' % (len(winners), len(rec['participants'])))
        for w in winners:
            print('  #%d %s -> %s' % (w['rank'], w['id'], w['prize']))

File: verifiers/draw-verify.js
/* ============================================================================
 * UVS verifiable allocation — REFERENCE VERIFIER (Node.js)
 *
 * The whole thing is ONE operation: a seeded random PERMUTATION of the
 * participants, then a published prize pool dealt onto that order.
 *
 *   combinedSeed = SHA-256( serverSeed + ":" + drandRandomness )
 *   score(id)    = SHA-256( combinedSeed + ":" + id )
 *   permutation  = participants sorted by score DESC  (ties: id ASC)
 *   allocation   = order[i] receives prizes[i]   (null beyond the pool)
 *
 * Single lookup: one participant's rank + prize, without ranking everyone.
 *
 * No dependencies beyond the standard library. Same algorithm in
 * draw_verify.py / DrawVerify.java / draw_verify.cpp — all reproduce
 * test-vectors.json byte-for-byte. Re-run any of them, or write your own.
 *
 * Run:  node draw-verify.js <record.json> [id]
 * ========================================================================== */
'use strict';
const crypto = require('crypto');
const sha256 = (s) => crypto.createHash('sha256').update(s, 'utf8').digest('hex');

function combinedSeed(serverSeed, drandRandomness) { return sha256(serverSeed + ':' + drandRandomness); }
function scoreOf(combined, id) { return sha256(combined + ':' + id); }

// participant ids MUST be unique (uvLs §3.1): with a duplicate the id tie-break no longer
// yields a total order and two entries collide on the same score. Reject, don't rank.
// (ids are assumed NFC-normalized by the producer, per §3.1.)
function requireUnique(participants) {
  if (new Set(participants).size !== participants.length)
    throw new Error('INVALID: duplicate participant ids — record rejected (uvLs §3.1)');
}

// deterministic order: highest score first; ties broken by id ascending
function cmp(a, b) {
  if (a.score > b.score) return -1;
  if (a.score < b.score) return 1;
  return a.id < b.id ? -1 : a.id > b.id ? 1 : 0;
}

function permute(participants, combined) {
  requireUnique(participants);
  return participants.map((id) => ({ id, score: scoreOf(combined, id) })).sort(cmp);
}

// full allocation: each position in the permutation gets prizes[i]
function allocate(participants, combined, prizes) {
  return permute(participants, combined).map((p, i) => ({ id: p.id, rank: i + 1, prize: i < prizes.length ? prizes[i] : null, score: p.score }));
}

// single lookup — O(M) hashing, no sort
function lookup(participants, combined, id, prizes) {
  requireUnique(participants);
  const me = scoreOf(combined, id);
  let higher = 0, present = false;
  for (const a of participants) {
    if (a === id) { present = true; continue; }
    const s = scoreOf(combined, a);
    if (s > me || (s === me && a < id)) higher++;
  }
  const rank = higher + 1;
  return { id, present, rank, prize: present && rank <= prizes.length ? prizes[rank - 1] : null, score: me };
}

// §6.1 proportional pools: a tier's winner count derived from the participant count M as an
// integer num/den with one rounding mode. All-integer (BigInt) so JS, Python, Java (long) and
// C++ (int64) agree; operands are non-negative, so division is floor (no truncation ambiguity).
function resolveCount(M, rule) {
  const num = rule.num, den = rule.den, mode = rule.mode || 'round-half-up';
  if (!Number.isInteger(num) || !Number.isInteger(den) || !Number.isInteger(M) || den <= 0 || num < 0 || M < 0)
    throw new Error('INVALID: proportional num/den/M must be non-negative integers with den>0 (uvLs §6.1)');
  const Mb = BigInt(M), n = BigInt(num), d = BigInt(den);
  let c;
  if (mode === 'floor') c = (Mb * n) / d;
  else if (mode === 'ceil') c = (Mb * n + d - 1n) / d;
  else if (mode === 'round-half-up') c = (2n * Mb * n + d) / (2n * d);
  else throw new Error('INVALID: unknown rounding mode "' + mode + '" (uvLs §6.1)');
  return Number(c);
}

// build the prize pool: explicit `prizes`; a §6 `prizePool` of {tier,count[,rule]} tiers
// (proportional tiers carry a §6.1 rule); or `winners` seats.
function poolOf(rec) {
  if (Array.isArray(rec.prizes)) return rec.prizes;
  if (rec.rules && Array.isArray(rec.rules.prizes)) return rec.rules.prizes;
  // a published draw record nests the pool under `rules` (uvLs §7); accept either shape.
  const pp = Array.isArray(rec.prizePool) ? rec.prizePool
           : (rec.rules && Array.isArray(rec.rules.prizePool) ? rec.rules.prizePool : null);
  if (pp) {
    const M = rec.participants.length, prizes = [];
    let total = 0;
    for (const t of pp) {
      if (typeof t.tier !== 'string') throw new Error('INVALID: tier label must be a string (uvLs §6)');
      let count;
      if (t.rule) {
        count = resolveCount(M, t.rule);
        // §6.1/§9.4: a proportional tier carries both the rule and its resolved count; reject a mismatch.
        if (t.count != null && t.count !== count)
          throw new Error('INVALID: tier "' + t.tier + '" count ' + t.count + ' != rule-resolved ' + count + ' (uvLs §6.1)');
      } else {
        if (!Number.isInteger(t.count)) throw new Error('INVALID: tier count must be a JSON integer (uvLs §6.1)');
        count = t.count;
      }
      if (total + count > M) count = M - total;   // §6.1 ordering: clamp running total to M
      for (let k = 0; k < count; k++) prizes.push(t.tier);
      total += count;
    }
    return prizes;
  }
  const r = rec.rules || {};
  const n = rec.winners || rec.N || r.winners || r.N || 0;
  return Array.from({ length: n }, () => rec.prizeLabel || r.prizeLabel || 'WIN');
}

// ── §5.4 anchor round rule (optional) ────────────────────────────────────────
// drand quicknet: 3s period, genesis 1692803367. A draw bound by the DERIVED-R rule (uvLs §5.4.1)
// sets R = roundAt(genTime)+1, so genTime < timeOfRound(R) holds by construction and the operator
// has no choice over R. Given the TSA token's genTime (from `openssl ts -reply -text`) and the
// record's round, confirm the binding. This checks ORDERING; verify the token itself with `openssl ts -verify`.
const QUICKNET = { genesis: 1692803367, period: 3 };
function roundAt(unixSec) { return Math.floor((unixSec - QUICKNET.genesis) / QUICKNET.period) + 1; }
function timeOfRound(round) { return QUICKNET.genesis + (round - 1) * QUICKNET.period; }
function checkAnchorRound(genTime, round) {
  const expectedRound = roundAt(genTime) + 1, roundTime = timeOfRound(round);
  return { ok: round === expectedRound && genTime < roundTime, expectedRound, roundTime, genBeforeRound: genTime < roundTime };
}

module.exports = { sha256, combinedSeed, scoreOf, permute, allocate, lookup, poolOf, resolveCount, roundAt, timeOfRound, checkAnchorRound };

// ---- CLI ----
if (require.main === module) {
  const fs = require('fs');
  const file = process.argv[2], id = process.argv[3];
  if (!file) { console.error('usage: node draw-verify.js <record.json> [id]'); process.exit(2); }
  const rec = JSON.parse(fs.readFileSync(file, 'utf8'));
  try { requireUnique(rec.participants); } catch (e) { console.error(e.message); process.exit(1); }
  const combined = combinedSeed(rec.serverSeed, rec.drand.randomness);
  const prizes = poolOf(rec);
  console.log('combinedSeed = SHA-256(serverSeed:drandRandomness) =', combined);
  const ca = rec.commitmentAnchor;
  if (ca && ca.genTime != null && rec.drand && rec.drand.round != null && String(ca.roundRule || '').startsWith('roundAt')) {
    const c = checkAnchorRound(ca.genTime, rec.drand.round);
    console.log('§5.4 derived-R: R==roundAt(genTime)+1 ?', rec.drand.round === c.expectedRound,
      '| genTime<timeOfRound(R) ?', c.genBeforeRound, '->', c.ok ? 'OK' : 'FAIL');
  }
  if (id) {
    const r = lookup(rec.participants, combined, id, prizes);
    console.log(`${id}: ${r.present ? 'rank ' + r.rank + ' of ' + rec.participants.length + ' -> ' + (r.prize || 'no prize') : 'NOT in the committed list'}`);
  } else {
    const a = allocate(rec.participants, combined, prizes);
    const winners = a.filter((x) => x.prize != null);
    console.log(`${winners.length} prize(s) among ${rec.participants.length} participants:`);
    winners.forEach((w) => console.log(`  #${w.rank} ${w.id} -> ${w.prize}`));
  }
}
